packages/cli/src/util/env/refresh-oidc-token.ts

```diff
@@ -0,0 +1,255 @@
+    if (!oidcToken) {
+      // If we fail to pull environment variables (for example, because we are
+      // temporarily offline), then we will continue trying once every minute
+      // until successful or aborted.
+      const envValues = await pullEnvValuesUntilSuccessful(
+        controller.signal,
+        client,
+        projectId,
+        source,
+        THROTTLE_MS
+      );
+      if (!envValues) {
+        return;
+      }
+      oidcToken = envValues[VERCEL_OIDC_TOKEN];
+      lastPulledEnvAt = clock();
+    }
+
+    if (!oidcToken) {
+      output.debug(
+        `${VERCEL_OIDC_TOKEN} is absent from environment variables; will not attempt to refresh it`
+      );
+      return;
+    }
+
+    const now = clock();
+
+    const exp = getExpFromOidcToken(oidcToken);
+    if (exp === null) {
+      output.debug(
+        `Cannot get "exp" claim from ${VERCEL_OIDC_TOKEN}; will not attempt to refresh it`
+      );
+      return;
+    }
+
+    const expiresAfterMs = exp * 1000 - now;
+    if (!Number.isFinite(expiresAfterMs)) {
+      output.debug(
+        `${VERCEL_OIDC_TOKEN} "exp" claim is invalid; will not attempt to refresh it`
+      );
+      return;
+    }
+
+    // If the OIDC token isn't already expired, patch .env.local.
+    // TODO(mroberts): Is this the only file we should care about?
+    const filename = '.env.local';
+    if (expiresAfterMs > 0) {
+      try {
+        await patchLocalEnv(client.cwd, filename, VERCEL_OIDC_TOKEN, oidcToken);
+      } catch (error) {
+        output.debug(`Failed to patch ${VERCEL_OIDC_TOKEN} in ${filename}`);
+      }
+      if (controller.signal.aborted) {
+        return;
+      }
+    } else {
+      output.debug(
+        `${VERCEL_OIDC_TOKEN} is already expired; skip writing to ${filename}`
+      );
+    }
+
+    // Schedule to refresh the OIDC token shortly before it expires.
+    let refreshAfterMs = Math.max(0, expiresAfterMs - REFRESH_BEFORE_EXPIRY_MS);
+
+    // Avoid invoking ourselves too frequently (wait at least one minute).
+    if (
+      lastPulledEnvAt !== undefined &&
+      now + refreshAfterMs - lastPulledEnvAt < THROTTLE_MS
+    ) {
+      refreshAfterMs = THROTTLE_MS;
+    }
+
+    if (expiresAfterMs < 0) {
+      output.debug(
+        `${VERCEL_OIDC_TOKEN} expired ${Math.abs(expiresAfterMs)} milliseconds ago; attempting to refresh it in ${refreshAfterMs} milliseconds`
+      );
+    } else {
+      output.debug(
+        `${VERCEL_OIDC_TOKEN} expires in ${expiresAfterMs} milliseconds; will attempt to refresh it in ${refreshAfterMs} milliseconds`
+      );
+    }
+
+    timeout = setTimeout(() => void go(), refreshAfterMs);
+  }
+
+  await go(oidcToken);
+
+  return () => {
+    controller.abort();
+    clearTimeout(timeout);
+  };
+}
+
+/**
+ * The OIDC token is a JSON Web Token (JWT). Decode the JWT and get its numeric
+ * "exp" claim, returning `null` on error.
+ */
+function getExpFromOidcToken(oidcToken: string): number | null {
```

Comment:
we can use `jose` instead of manual decoding, see https://github.com/vercel/vercel/blob/5c7f57fdd9339715e186c0281f52d989ca8bbb13/packages/cli/src/util/oauth.ts#L387

Reply:
Good idea. I used `decodeJwt` from jose directly, since I don't think we need to validate.